core/src/test/java/nl/nn/adapterframework/pipes/PgpPipeTest.java

```diff
@@ -0,0 +1,203 @@
+				{"Verify Someone Signed", "success",
+						new String[]{"sign", sender[2], sender[1], sender[4], sender[0], recipient[0]},
+						new String[]{"verify", recipient[2], recipient[1], recipient[4], null, recipient[0]}},
+				{"Encrypt then Verify", "org.bouncycastle.openpgp.PGPException",
+						new String[]{"encrypt", null, null, recipient[3], null, recipient[0]},
+						new String[]{"verify", recipient[2], recipient[1], recipient[4], sender[0], recipient[0]}},
+				{"Sign wrong params", "nl.nn.adapterframework.configuration.ConfigurationException",
+						new String[]{"sign", null, null, recipient[3], null, recipient[0]},
+						new String[]{"decrypt", recipient[2], recipient[1], null, null, null}},
+				{"Null action", "nl.nn.adapterframework.configuration.ConfigurationException",
+						new String[]{null, null, null, recipient[3], null, recipient[0]},
+						new String[]{"decrypt", recipient[2], recipient[1], null, null, null}},
+				{"Non-existing action", "nl.nn.adapterframework.configuration.ConfigurationException",
+						new String[]{"non-existing action", null, null, recipient[3], null, recipient[0]},
+						new String[]{"decrypt", recipient[2], recipient[1], null, null, null}},
+				{"Wrong password", "org.bouncycastle.openpgp.PGPException",
+						new String[]{"encrypt", null, null, recipient[3], null, recipient[0]},
+						new String[]{"decrypt", recipient[2], "wrong password :/", null, null, null}},
+				{"Decrypt Plaintext", "org.bouncycastle.openpgp.PGPException",
+						new String[]{"decrypt", recipient[2], recipient[1], null, null, null},
+						new String[]{"decrypt", recipient[2], recipient[1], null, null, null}},
+		});
+	}
+
+	public PgpPipeTest(String name, String expectation, String[] encryptParams, String[] decryptParams) {
+		setup();
+		this.expectation = expectation;
+		this.encryptParams = encryptParams;
+		this.decryptParams = decryptParams;
+	}
+
+	@Test
+	public void dotest() throws Throwable {
+		try {
+			// Configure pipes
+			configurePipe(encryptPipe, encryptParams);
+			configurePipe(decryptPipe, decryptParams);
+
+			// Encryption phase
+			PipeRunResult encryptionResult = encryptPipe.doPipe(MESSAGE, session);
+
+			// Make sure it's PGP message
+			OutputStream mid = (OutputStream) encryptionResult.getResult();
+			assertMessage(mid.toString(), MESSAGE);
+
+			// Decryption phase
+			PipeRunResult decryptionResult = decryptPipe.doPipe(encryptionResult.getResult(), session);
+			OutputStream result = (OutputStream) decryptionResult.getResult();
+
+			// Assert decrypted message equals to the original message
+			Assert.assertEquals(MESSAGE, result.toString());
+			Assert.assertEquals("success", expectation);
+		} catch (Exception e) {
+			if (checkExceptionClass(e, expectation)) {
+				Assert.assertTrue(true);
+			} else {
+				throw e;
+			}
+		}
+	}
+
+	/**
+	 * Creates pipes and pipeline session base for testing.
+	 */
+	public void setup() {
```

Comment:
Wont it be easier to use the Before annotation?

Reply:
It does not matter from execution perspective. But yeah I guess @before is cleaner.